### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [41]:
# The best way to determine the number of players is to look at unique screen names

num_players = purchase_data.SN.nunique()

# Now create a dataframe using that data

data = {"Total Players": [num_players]}
playerDF = pd.DataFrame(data)

playerDF

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [73]:
# The way I've chosen to attack the rest of these questions is by building series and then dataframes from
# those series.

# Create pandas series for unique items in Item ID, and then name the series

unique_items = pd.Series(purchase_data["Item ID"].nunique(), name = "Number of Unique Items")


# Create series finding the mean price, naming the series, and then rounding to two places

avg_price = pd.Series(purchase_data.Price.mean(), name = "Average Price")
avg_price = round(avg_price, 2)


# Create series for unique purchases, and name it

num_purchases = pd.Series(purchase_data["Purchase ID"].nunique(), name = "Number of Purchases")


# Create series that sums the prices, and name it

total_rev = pd.Series(purchase_data.Price.sum(), name = "Total Revenue")


# Concatenate the series, then convert them into pandas dataframe

df = pd.concat([unique_items, avg_price, num_purchases, total_rev], axis = 1)
Purchase_AnalysisDF = pd.DataFrame(df)
Purchase_AnalysisDF

# Making the formatting pretty (including hiding the index)

format_dict = {'Average Price':'${0:,.2f}', 'Total Revenue': '${0:,.2f}'}
Purchase_AnalysisDF.style.format(format_dict)

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [50]:
# Find number of unique users by gender and then make numpy array for calculation of percentages

gen = purchase_data.groupby(["Gender"])["SN"].nunique()
np.array(gen)

# Making a new array of percentages 

pct_gen = gen/num_players

# Turning the arrays into panda series with names

s1 = pd.Series(gen, name = "Total Count") 
s2 = pd.Series(pct_gen, name = "Percentage of Players")

# Concatenating the series to make a new dataframe

newDF = pd.concat([s1, s2], axis = 1)

# Making the formatting pretty

format_dict = {"Percentage of Players": '{:.2%}'}
newDF.style.format(formatdict)



## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [74]:
# Create series of purchases by gender, then count

pur_count = pd.Series(purchase_data.groupby(["Gender"])["Purchase ID"].count(), name = "Purchase Count")


# Create series of the prices paid by gender

tot_pur_val = pd.Series(purchase_data.groupby(["Gender"])["Price"].agg(sum), name = "Total Purchase Value")


# Create series of average purchase price and round to two places

avg_pur_price = pd.Series(round(tot_pur_val/pur_count, 2), name = "Average Purchase Price")


# Divide total price paid by the total count of players by gender to get average total per person

avg_per_person = pd.Series(round(tot_pur_val/gen, 2), name = "Avg Total Purchase per Person")

# Concatenate the series, then turn into a dataframe

pur_analysisPD = pd.concat([pur_count, avg_pur_price, tot_pur_val, avg_per_person], axis = 1)
pur_analysisPD = pd.DataFrame(pur_analysisPD)
pur_analysisPD

format_dict = {'Average Purchase Price':'${0:,.2f}', 'Total Purchase Value': '${0:,.2f}',
               'Avg Total Purchase per Person': '${0:,.2f}'}
pur_analysisPD.style.format(format_dict)

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [56]:
# Create bins and name them

bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
bins_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]


# Pull out Age and assign to bins

purchase_data["Age Group"] = pd.cut(purchase_data["Age"], bins, labels = bins_names)

# Turn our binned data into a series

Age_Count = pd.Series(purchase_data.groupby("Age Group")["SN"].nunique(), name = "Total Count")


# Make a series of binned data as percentage

age_percentage = pd.Series(Age_Count/num_players, name = "Percentage of Players")

# Concatenate series and make into a dataframe

binned_ageDF = pd.concat([Age_Count, age_percentage], axis = 1)
binned_ageDF = pd.DataFrame(binned_ageDF)

# Making the formatting pretty
format_dict = {"Percentage of Players": '{:.2%}'}
binned_ageDF = binned_ageDF.style.format(format_dict)

binned_ageDF

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [75]:
# Since purchase_data already contains Age Groups thanks to the cell above, I can work with that column.
# I chose to leave the ages in order rather than move the under-10's to the bottom row.

# Series of unique purchases by age

age_pur_count = pd.Series(purchase_data.groupby("Age Group")["Purchase ID"].nunique(), name = "Purchase Count")

# Series of average purchase price by group

age_avg_price = pd.Series(round(purchase_data.groupby("Age Group")["Price"].mean(), 2), 
                          name = "Average Purchase Price")

# Series of the total price paid by group

age_tot_val = pd.Series(purchase_data.groupby("Age Group")["Price"].sum(), name = "Total Purchase Value")

# Series of the average total per person, using Age_Count from cell above

age_per_person = pd.Series(round(age_tot_val/Age_Count, 2), name = "Avg Total Purchase per Person")

# Concatenate and turn into dataframe

Analysis_by_AgeDF = pd.concat([age_pur_count, age_avg_price, age_tot_val, age_per_person], axis = 1)
Analysis_by_AgeDF = pd.DataFrame(Analysis_by_AgeDF)

# Making the formatting pretty

format_dict = {'Average Purchase Price':'${0:,.2f}', 'Total Purchase Value': '${0:,.2f}',
               'Avg Total Purchase per Person': '${0:,.2f}'}
Analysis_by_AgeDF = Analysis_by_AgeDF.style.format(format_dict)

Analysis_by_AgeDF

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [72]:
# After much hair-pulling and google-fu, I figured out how to do a combination of groupby and agg to 
# knock out a working dataframe in a single line. The count, mean, and sum functions within agg
# both name the column and perform the function named.

Top_Spender_df = pd.DataFrame(purchase_data.groupby("SN")["Price"].agg(["count", "mean","sum"]))

# Sort by purchase count

Top_Spender_df = Top_Spender_df.sort_values("count", ascending = False)

# Rename the columns that agg named

columns = {"count": "Purchase Count", "mean": "Average Purchase Price", "sum": "Total Purchase Value"}
Top_Spender_df = Top_Spender_df.rename(columns = columns)

# Making the formatting pretty doesn't seem to work here

# format_dict = {'Average Purchase Price':'${0:,.2f}', 'Total Purchase Value': '${0:,.2f}'}
# Top_Spender_df = Top_Spender_df.style.format(format_dict).hide_index()


# Clean up the formatting on purchase price
Top_Spender_df["Average Purchase Price"] = round(Top_Spender_df["Average Purchase Price"],2)


# Top 5 in descending order

Top_Spender_df.head()

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                    3.79                 18.96
Iral74                    4                    3.40                 13.62
Idastidru52               4                    3.86                 15.45
Asur53                    3                    2.48                  7.44
Inguron55                 3                    3.70                 11.11

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [11]:
# Now I'll do groupby/agg using Item ID and Item Name as top level and then will do counts, mean, and sum
# based on price.

Most_Pop_df = pd.DataFrame(purchase_data.groupby(["Item ID", "Item Name"])["Price"].agg(["count", "mean","sum"]))

# Again the sort will be on count, since we're finding out which game is most popular.

Most_Pop_df = Most_Pop_df.sort_values("count", ascending = False)

# Renaming the columns.

columns = {"count": "Purchase Count", "mean": "Item Price", "sum": "Total Purchase Value"}
Most_Pop_df = Most_Pop_df.rename(columns = columns)

# Top 5 in descending order
    
Most_Pop_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                      Item Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
145     Fiery Glass Crusader                                4.58   
108     Extraction, Quickblade Of Trembling Hands           3.53   
82      Nirvana                                             4.90   
19      Pursuit, Cudgel of Necromancy                       1.02   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
108     Extraction, Quickblade Of Trembling Hands                    31.77  
82      Nirvana                                                      44.10  
19      Pursuit, Cudgel of Necromancy                                 8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
# This dataframe is exactly the same as Most_Pop_df

Most_Profit_df = pd.DataFrame(purchase_data.groupby(["Item ID", "Item Name"])["Price"].agg(["count", "mean","sum"]))

# The difference is that it's sorted by Total Purchase Value (which has yet to be renamed in this line)

Most_Profit_df = Most_Profit_df.sort_values("sum", ascending = False)

# Renaming the columns
columns = {"count": "Purchase Count", "mean": "Item Price", "sum": "Total Purchase Value"}
Most_Profit_df = Most_Profit_df.rename(columns = columns)

# Top 5 in descending order

Most_Profit_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                      Item Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
82      Nirvana                                             4.90   
145     Fiery Glass Crusader                                4.58   
92      Final Critic                                        4.88   
103     Singed Scalpel                                      4.35   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
92      Final Critic                                                 39.04  
103     Singed Scalpel                                               34.80





Three trends I see:

1) Games that are purchased most often are (not surprisingly) the most profitable or nearly the most profitable.

2) On a per-person average by gender, the best spenders are female. Out of curiosity, I checked the gender of the top five spenders overall. They're all male.

3) By age, kids under 10 and adults aged 35-39 (who also happen to be the right age to have kids under 10) spend the most per person. It might be worth trying to develop games that appeal to them -- but it's more of a sure thing to stick with the core group of 20-24-year-olds. They account for 44% of gamers and 48% ($1114/$2380) of total revenue.
